File: packages/qa_native/src/qa_audio_device.c
// The output device and the transport (audio program 2C).
//
// This is where "audio is the master clock" becomes literal. The device
// pulls buffers on its own realtime thread; each pull mixes the next span
// of samples and advances a counter. The playback position IS that counter
// — the number of samples handed to the device — so it cannot drift from
// what is being heard. A free-running timer can; that is the defect this
// program started from.
//
// NON-NEGOTIABLE: no Dart runs in the callback. Dart uploads the schedule
// and the PCM, starts and stops the transport, and polls the position.
// Everything the callback touches is already resident C memory.
//
// The schedule is only replaced while the transport is STOPPED. That makes
// the handoff trivially safe without lock-free machinery, and it matches
// how playback already works — the schedule is built when playback starts.
// Editing sound during playback would need a double-buffered swap; that is
// a later problem, and pretending otherwise would put a data race on the
// realtime thread.

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#if defined(_WIN32)
#define QA_EXPORT __declspec(dllexport)
#else
#define QA_EXPORT __attribute__((visibility("default")))
#endif

// Playback only: no capture, no decoding (conforms are already PCM), no
// resampling (imports already landed at the project rate), no generation.
// Each of those is code we would ship and never run.
#define MA_NO_ENCODING
#define MA_NO_DECODING
#define MA_NO_GENERATION
#define MA_NO_RESOURCE_MANAGER
#define MA_NO_NODE_GRAPH
#define MA_NO_ENGINE
#define MA_IMPLEMENTATION
#include "third_party/miniaudio/miniaudio.h"

// The mixer's types live in qa_engine.c; declared here rather than shared
// through a header because the project keeps one portable C file per
// concern and no headers of its own.
typedef struct {
  double gain;
  int64_t start_sample;
  int64_t end_sample;
  int64_t source_offset;
  int64_t fade_in_samples;
  int64_t fade_out_samples;
  int32_t source_index;
  int32_t reserved;
} qa_audio_clip;

typedef struct {
  int64_t source_start;
  int64_t length;
  int32_t channels;
  int32_t reserved;
  const float* samples;
} qa_audio_source;

extern void qa_audio_mix(const qa_audio_clip* clips,
                         int32_t clip_count,
                         const qa_audio_source* sources,
                         int32_t source_count,
                         int64_t start_sample,
                         int32_t sample_count,
                         int32_t out_channels,
                         double* out);

#define QA_DEVICE_MAX_BLOCK 8192

typedef struct {
  ma_device device;
  int32_t device_open;
  int32_t playing;

  // The transport. `position` counts samples HANDED TO THE DEVICE, which
  // is what makes it the clock: it advances only when audio actually
  // leaves, so it cannot run ahead of what is heard.
  int64_t position;
  int64_t start_position;
  int64_t stop_position;  // exclusive; <= start means "no end"
  int32_t looping;

  qa_audio_clip* clips;
  int32_t clip_count;
  qa_audio_source* sources;
  int32_t source_count;
  float* source_pcm;  // one owned block holding every source's samples
  int64_t source_pcm_floats;

  int32_t channels;
  int32_t sample_rate;
  double* scratch;  // the mix bus, sized once at open
} qa_audio_device_state;

static qa_audio_device_state g_audio;

// The realtime callback. memcpy and a weighted sum, nothing else: no
// allocation, no locks, no I/O, no Dart.
static void qa_audio_data_callback(ma_device* device,
                                   void* output,
                                   const void* input,
                                   ma_uint32 frame_count) {
  (void)device;
  (void)input;
  float* out = (float*)output;
  const int32_t channels = g_audio.channels;
  const size_t total = (size_t)frame_count * (size_t)channels;
  memset(out, 0, total * sizeof(float));

  if (!g_audio.playing || g_audio.scratch == NULL) {
    return;
  }

  ma_uint32 done = 0;
  while (done < frame_count) {
    ma_uint32 block = frame_count - done;
    if (block > QA_DEVICE_MAX_BLOCK) {
      block = QA_DEVICE_MAX_BLOCK;
    }

    int64_t position = g_audio.position;
    if (g_audio.stop_position > g_audio.start_position &&
        position >= g_audio.stop_position) {
      if (g_audio.looping) {
        position = g_audio.start_position;
        g_audio.position = position;
      } else {
        g_audio.playing = 0;
        return;
      }
    }
    // Never mix past the stop point in one block; the wrap has to land on
    // the sample, not the buffer boundary.
    if (g_audio.stop_position > g_audio.start_position) {
      const int64_t remaining = g_audio.stop_position - position;
      if ((int64_t)block > remaining) {
        block = (ma_uint32)remaining;
      }
    }
    if (block == 0) {
      return;
    }

    qa_audio_mix(g_audio.clips, g_audio.clip_count, g_audio.sources,
                 g_audio.source_count, position, (int32_t)block, channels,
                 g_audio.scratch);

    const size_t written = (size_t)block * (size_t)channels;
    float* target = out + (size_t)done * (size_t)channels;
    for (size_t index = 0; index < written; index += 1) {
      double value = g_audio.scratch[index];
      // The bus has headroom; the DEVICE does not. Clipping belongs here,
      // at the boundary, exactly as it does for the int16 path.
      if (value > 1.0) {
        value = 1.0;
      } else if (value < -1.0) {
        value = -1.0;
      }
      target[index] = (float)value;
    }

    g_audio.position = position + (int64_t)block;
    done += block;
  }
}

static void qa_audio_free_schedule(void) {
  free(g_audio.clips);
  free(g_audio.sources);
  free(g_audio.source_pcm);
  g_audio.clips = NULL;
  g_audio.sources = NULL;
  g_audio.source_pcm = NULL;
  g_audio.clip_count = 0;
  g_audio.source_count = 0;
  g_audio.source_pcm_floats = 0;
}

// Opens the output device. [backend] 0 = let miniaudio choose, 1 = the
// NULL backend, which runs the real callback on a real thread with no
// hardware — how the transport is tested on a CI runner with no sound card.
//
// Returns the device's actual sample rate, or 0 on failure. The rate is
// returned rather than assumed because a device may refuse the one asked
// for, and conforming to the wrong rate would put every sound at the wrong
// speed.
QA_EXPORT int32_t qa_audio_device_open(int32_t sample_rate,
                                       int32_t channels,
                                       int32_t backend) {
  if (g_audio.device_open) {
    return g_audio.sample_rate;
  }
  if (sample_rate <= 0 || channels <= 0 || channels > 8) {
    return 0;
  }

  ma_device_config config = ma_device_config_init(ma_device_type_playback);
  config.playback.format = ma_format_f32;
  config.playback.channels = (ma_uint32)channels;
  config.sampleRate = (ma_uint32)sample_rate;
  config.dataCallback = qa_audio_data_callback;

  ma_result result;
  if (backend == 1) {
    ma_backend backends[1];
    backends[0] = ma_backend_null;
    static ma_context context;
    if (ma_context_init(backends, 1, NULL, &context) != MA_SUCCESS) {
      return 0;
    }
    result = ma_device_init(&context, &config, &g_audio.device);
  } else {
    result = ma_device_init(NULL, &config, &g_audio.device);
  }
  if (result != MA_SUCCESS) {
    return 0;
  }

  g_audio.channels = (int32_t)g_audio.device.playback.channels;
  g_audio.sample_rate = (int32_t)g_audio.device.sampleRate;
  g_audio.scratch = (double*)malloc((size_t)QA_DEVICE_MAX_BLOCK *
                                    (size_t)g_audio.channels * sizeof(double));
  if (g_audio.scratch == NULL) {
    ma_device_uninit(&g_audio.device);
    return 0;
  }
  g_audio.device_open = 1;
  g_audio.playing = 0;
  g_audio.position = 0;

  if (ma_device_start(&g_audio.device) != MA_SUCCESS) {
    free(g_audio.scratch);
    g_audio.scratch = NULL;
    ma_device_uninit(&g_audio.device);
    g_audio.device_open = 0;
    return 0;
  }
  return g_audio.sample_rate;
}

QA_EXPORT void qa_audio_device_close(void) {
  if (!g_audio.device_open) {
    return;
  }
  g_audio.playing = 0;
  ma_device_uninit(&g_audio.device);
  free(g_audio.scratch);
  g_audio.scratch = NULL;
  qa_audio_free_schedule();
  g_audio.device_open = 0;
}

QA_EXPORT int32_t qa_audio_device_is_open(void) { return g_audio.device_open; }
QA_EXPORT int32_t qa_audio_device_sample_rate(void) { return g_audio.sample_rate; }
QA_EXPORT int32_t qa_audio_device_channels(void) { return g_audio.channels; }

// The device's own reported output latency in samples — what the picture
// has to be pulled forward by so it matches what is being heard. Whatever
// this cannot account for is the residual the user's A/V offset removes.
QA_EXPORT int64_t qa_audio_device_latency_samples(void) {
  if (!g_audio.device_open) {
    return 0;
  }
  return (int64_t)g_audio.device.playback.internalPeriodSizeInFrames *
         (int64_t)g_audio.device.playback.internalPeriods;
}

// Replaces the schedule. Only legal while STOPPED (see the file header) —
// returns 0 and changes nothing otherwise, rather than racing the
// realtime thread.
//
// The PCM is COPIED into one owned block: the callback must never chase a
// pointer into Dart-managed memory that could move or be collected.
QA_EXPORT int32_t qa_audio_device_set_schedule(
    const qa_audio_clip* clips,
    int32_t clip_count,
    const qa_audio_source* sources,
    int32_t source_count,
    const float* pcm,
    int64_t pcm_floats,
    const int64_t* source_offsets) {
  if (g_audio.playing) {
    return 0;
  }
  qa_audio_free_schedule();
  if (clip_count < 0 || source_count < 0 || pcm_floats < 0) {
    return 0;
  }

  if (clip_count > 0) {
    g_audio.clips =
        (qa_audio_clip*)malloc((size_t)clip_count * sizeof(qa_audio_clip));
    if (g_audio.clips == NULL) {
      return 0;
    }
    memcpy(g_audio.clips, clips, (size_t)clip_count * sizeof(qa_audio_clip));
    g_audio.clip_count = clip_count;
  }

  if (pcm_floats > 0) {
    g_audio.source_pcm = (float*)malloc((size_t)pcm_floats * sizeof(float));
    if (g_audio.source_pcm == NULL) {
      qa_audio_free_schedule();
      return 0;
    }
    memcpy(g_audio.source_pcm, pcm, (size_t)pcm_floats * sizeof(float));
    g_audio.source_pcm_floats = pcm_floats;
  }

  if (source_count > 0) {
    g_audio.sources =
        (qa_audio_source*)malloc((size_t)source_count * sizeof(qa_audio_source));
    if (g_audio.sources == NULL) {
      qa_audio_free_schedule();
      return 0;
    }
    memcpy(g_audio.sources, sources,
           (size_t)source_count * sizeof(qa_audio_source));
    // Repoint every source at our copy: the caller's `samples` pointers
    // describe ITS memory, and the offsets say where each block landed in
    // the flattened PCM.
    for (int32_t index = 0; index < source_count; index += 1) {
      const int64_t offset = source_offsets == NULL ? 0 : source_offsets[index];
      g_audio.sources[index].samples =
          (offset >= 0 && offset < pcm_floats) ? g_audio.source_pcm + offset
                                               : NULL;
    }
    g_audio.source_count = source_count;
  }
  return 1;
}

// Starts playback at [start_sample]. [stop_sample] is exclusive; pass a
// value at or below the start for "no end".
QA_EXPORT int32_t qa_audio_device_play(int64_t start_sample,
                                       int64_t stop_sample,
                                       int32_t looping) {
  if (!g_audio.device_open) {
    return 0;
  }
  g_audio.position = start_sample;
  g_audio.start_position = start_sample;
  g_audio.stop_position = stop_sample;
  g_audio.looping = looping ? 1 : 0;
  g_audio.playing = 1;
  return 1;
}

QA_EXPORT void qa_audio_device_stop(void) { g_audio.playing = 0; }

QA_EXPORT int32_t qa_audio_device_is_playing(void) { return g_audio.playing; }

// Samples handed to the device so far — THE clock. The picture reads this
// and shows whatever frame it lands in; if rendering fell behind, frames
// are dropped rather than the sound being made to wait.
QA_EXPORT int64_t qa_audio_device_position(void) { return g_audio.position; }

// Moves the transport without restarting anything: because the mixer
// builds a mix rather than starting clips, seeking is just changing where
// the next block is read from.
QA_EXPORT void qa_audio_device_seek(int64_t sample) {
  g_audio.position = sample;
}

File: lib/src/native/qa_audio_device.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import '../services/audio/audio_mixer_reference.dart';
import 'qa_audio_native.dart';
import 'qa_engine_abi.dart';

/// The output device and the transport (audio program 2C).
///
/// This is where "audio is the master clock" stops being a design note.
/// [positionSamples] counts samples HANDED TO THE DEVICE, so it advances
/// only when audio actually leaves — it cannot run ahead of what is heard,
/// which a free-running timer can and does.
///
/// Dart's whole job here is upload, transport control, and polling. No Dart
/// runs in the callback; that is the rule the realtime thread is built on.
///
/// Absence is graceful: no binary, or a device that refuses to open,
/// leaves [isOpen] false and the caller on its existing path. Silence is
/// never an acceptable outcome for audio.
final class QaAudioDevice {
  QaAudioDevice._(this._library);

  final DynamicLibrary _library;

  static QaAudioDevice? _instance;
  static bool _tried = false;

  /// Test hook: point the loader at a locally built binary.
  static String? debugLibraryPathOverride;

  static void debugResetForTests() {
    _instance = null;
    _tried = false;
  }

  /// The device, or null when no binary loads, the binary speaks a
  /// different ABI, or it disagrees about the audio struct layout.
  ///
  /// That last check matters more here than anywhere else: this class
  /// hands [QaAudioClipStruct] and friends — declared next door in
  /// `qa_audio_native.dart` — straight to the C on the path that reaches
  /// a real speaker. Until the AUDIO-PRO ABI consolidation it was the one
  /// loader with NO gate at all, so a layout change would have stood the
  /// verification mixer down correctly while the device kept pushing
  /// misread fields at someone's headphones.
  static QaAudioDevice? get instance {
    if (!_tried) {
      _tried = true;
      final library = openQaEngineLibrary(
        overridePath: debugLibraryPathOverride,
      );
      _instance = library == null || !qaAudioStructLayoutsMatch(library)
          ? null
          : QaAudioDevice._(library);
    }
    return _instance;
  }

  late final _open = _library
      .lookupFunction<
        Int32 Function(Int32, Int32, Int32, Int32),
        int Function(int, int, int, int)
      >('qa_audio_device_open');
  late final _deviceCount = _library
      .lookupFunction<Int32 Function(Int32, Int32), int Function(int, int)>(
        'qa_audio_device_count',
      );
  late final _deviceDescribe = _library.lookupFunction<
    Int32 Function(Int32, Int32, Pointer<Utf8>, Int32, Pointer<Int32>),
    int Function(int, int, Pointer<Utf8>, int, Pointer<Int32>)
  >('qa_audio_device_describe');
  late final _close = _library
      .lookupFunction<Void Function(), void Function()>('qa_audio_device_close');
  late final _isOpen = _library
      .lookupFunction<Int32 Function(), int Function()>('qa_audio_device_is_open');
  late final _sampleRate = _library
      .lookupFunction<Int32 Function(), int Function()>(
        'qa_audio_device_sample_rate',
      );
  late final _channels = _library
      .lookupFunction<Int32 Function(), int Function()>(
        'qa_audio_device_channels',
      );
  late final _latency = _library
      .lookupFunction<Int64 Function(), int Function()>(
        'qa_audio_device_latency_samples',
      );
  late final _setSchedule = _library.lookupFunction<
    Int32 Function(
      Pointer<QaAudioClipStruct>,
      Int32,
      Pointer<QaAudioSourceStruct>,
      Int32,
      Pointer<Float>,
      Int64,
      Pointer<Int64>,
      Pointer<QaAudioEnvelopeKeyStruct>,
      Int32,
    ),
    int Function(
      Pointer<QaAudioClipStruct>,
      int,
      Pointer<QaAudioSourceStruct>,
      int,
      Pointer<Float>,
      int,
      Pointer<Int64>,
      Pointer<QaAudioEnvelopeKeyStruct>,
      int,
    )
  >('qa_audio_device_set_schedule');
  late final _play = _library
      .lookupFunction<Int32 Function(Int64, Int64, Int32), int Function(int, int, int)>(
        'qa_audio_device_play',
      );
  late final _stop = _library
      .lookupFunction<Void Function(), void Function()>('qa_audio_device_stop');
  late final _isPlaying = _library
      .lookupFunction<Int32 Function(), int Function()>(
        'qa_audio_device_is_playing',
      );
  late final _position = _library
      .lookupFunction<Int64 Function(), int Function()>(
        'qa_audio_device_position',
      );
  late final _seek = _library
      .lookupFunction<Void Function(Int64), void Function(int)>(
        'qa_audio_device_seek',
      );
  late final _peak = _library
      .lookupFunction<Double Function(Int32), double Function(int)>(
        'qa_audio_device_peak',
      );
  late final _captureStart = _library
      .lookupFunction<
        Int32 Function(Int32, Int32, Int32),
        int Function(int, int, int)
      >('qa_audio_capture_start');
  late final _captureRead = _library
      .lookupFunction<
        Int32 Function(Pointer<Float>, Int32),
        int Function(Pointer<Float>, int)
      >('qa_audio_capture_read');
  late final _captureStop = _library
      .lookupFunction<Void Function(), void Function()>('qa_audio_capture_stop');
  late final _captureIsOpen = _library
      .lookupFunction<Int32 Function(), int Function()>(
        'qa_audio_capture_is_open',
      );
  late final _captureChannels = _library
      .lookupFunction<Int32 Function(), int Function()>(
        'qa_audio_capture_channels',
      );
  late final _captureDroppedFrames = _library
      .lookupFunction<Int64 Function(), int Function()>(
        'qa_audio_capture_dropped_frames',
      );
  late final _captureLatency = _library
      .lookupFunction<Int64 Function(), int Function()>(
        'qa_audio_capture_latency_samples',
      );

  /// Opens the output device, returning its ACTUAL sample rate (0 = the
  /// device could not be opened).
  ///
  /// The rate comes back rather than being assumed because a device may
  /// refuse the one asked for, and conforming to a rate the device is not
  /// running at would put every sound at the wrong speed.
  ///
  /// [useNullBackend] runs miniaudio's null device: a real callback on a
  /// real thread with no hardware. That is how the transport is exercised
  /// on a CI runner with no sound card.
  ///
  /// [deviceIndex] picks from the last [devicesOf] enumeration (AUDIO-PRO
  /// R4); -1 = the system default. A bad index fails rather than opening
  /// something else — retry with -1 to fall back deliberately.
  int open({
    int sampleRate = 48000,
    int channels = 2,
    bool useNullBackend = false,
    int deviceIndex = -1,
  }) => _open(sampleRate, channels, useNullBackend ? 1 : 0, deviceIndex);

  /// Enumerates output ([capture] false) or input devices, in the index
  /// order [open]'s `deviceIndex` refers to. Empty on any failure —
  /// callers fall back to the system default, never crash a picker.
  List<({String name, bool isDefault})> devicesOf({
    required bool capture,
    bool useNullBackend = false,
  }) {
    final kind = capture ? 1 : 0;
    final count = _deviceCount(kind, useNullBackend ? 1 : 0);
    if (count <= 0) {
      return const [];
    }
    final nameBuffer = calloc<Uint8>(256);
    final isDefaultOut = calloc<Int32>();
    try {
      final devices = <({String name, bool isDefault})>[];
      for (var index = 0; index < count; index += 1) {
        final length = _deviceDescribe(
          kind,
          index,
          nameBuffer.cast<Utf8>(),
          256,
          isDefaultOut,
        );
        if (length < 0) {
          continue;
        }
        devices.add(
          (
            name: nameBuffer.cast<Utf8>().toDartString(),
            isDefault: isDefaultOut.value != 0,
          ),
        );
      }
      return devices;
    } finally {
      calloc.free(isDefaultOut);
      calloc.free(nameBuffer);
    }
  }

  void close() => _close();
  bool get isOpen => _isOpen() != 0;
  int get sampleRate => _sampleRate();
  int get channels => _channels();

  /// The device's reported output latency in samples — how far ahead of
  /// what is heard the buffer runs, and therefore how much the picture has
  /// to be pulled forward. What this cannot account for is the residual
  /// the user's A/V offset removes.
  int get latencySamples => _latency();

  /// Uploads the schedule and its PCM — legal at ANY time (AUDIO-PRO R3):
  /// the C builds the replacement in a standby slot and flips atomically,
  /// so a swap during playback is heard within one mixed block, with the
  /// old arrays freed only after the callback provably abandoned them.
  ///
  /// The PCM is flattened into one block and COPIED on the C side: the
  /// callback must never chase a pointer into Dart-managed memory.
  bool setSchedule({
    required List<AudioMixClip> clips,
    required List<AudioMixSource> sources,
  }) {
    var totalFloats = 0;
    final offsets = <int>[];
    for (final source in sources) {
      offsets.add(totalFloats);
      totalFloats += source.samples.length;
    }

    final clipArray = calloc<QaAudioClipStruct>(clips.isEmpty ? 1 : clips.length);
    final sourceArray = calloc<QaAudioSourceStruct>(
      sources.isEmpty ? 1 : sources.length,
    );
    final pcm = calloc<Float>(totalFloats <= 0 ? 1 : totalFloats);
    final offsetArray = calloc<Int64>(offsets.isEmpty ? 1 : offsets.length);
    // The clips' envelopes flatten into one shared key array, exactly as
    // the mixer FFI does (the C copies it beside the PCM).
    var envelopeTotal = 0;
    for (final clip in clips) {
      envelopeTotal += clip.envelope.length;
    }
    final envelopeArray = calloc<QaAudioEnvelopeKeyStruct>(
      envelopeTotal <= 0 ? 1 : envelopeTotal,
    );
    try {
      var envelopeCursor = 0;
      for (var index = 0; index < clips.length; index += 1) {
        final clip = clips[index];
        final target = clipArray[index];
        target.gain = clip.gain;
        target.panLeft = clip.panLeft;
        target.panRight = clip.panRight;
        target.startSample = clip.startSample;
        target.endSample = clip.endSample;
        target.sourceOffset = clip.sourceOffset;
        target.fadeInSamples = clip.fadeInSamples;
        target.fadeOutSamples = clip.fadeOutSamples;
        target.sourceIndex = clip.sourceIndex;
        target.fadeCurve = clip.fadeCurve;
        target.envelopeOffset = envelopeCursor;
        target.envelopeCount = clip.envelope.length;
        for (final point in clip.envelope) {
          final key = envelopeArray[envelopeCursor];
          key.sample = point.sample;
          key.gain = point.gain;
          envelopeCursor += 1;
        }
      }
      for (var index = 0; index < sources.length; index += 1) {
        final source = sources[index];
        final target = sourceArray[index];
        target.sourceStart = source.sourceStart;
        target.length = source.length;
        target.channels = source.channels;
        target.reserved = 0;
        target.samples = nullptr; // repointed at the C copy
        offsetArray[index] = offsets[index];
        if (source.samples.isNotEmpty) {
          pcm
              .asTypedList(totalFloats)
              .setRange(
                offsets[index],
                offsets[index] + source.samples.length,
                source.samples,
              );
        }
      }
      return _setSchedule(
            clipArray,
            clips.length,
            sourceArray,
            sources.length,
            pcm,
            totalFloats,
            offsetArray,
            envelopeArray,
            envelopeTotal,
          ) !=
          0;
    } finally {
      calloc.free(envelopeArray);
      calloc.free(offsetArray);
      calloc.free(pcm);
      calloc.free(sourceArray);
      calloc.free(clipArray);
    }
  }

  /// Starts at [startSample]. [stopSample] is exclusive; pass a value at
  /// or below the start for "no end".
  bool play({
    required int startSample,
    int stopSample = -1,
    bool looping = false,
  }) => _play(startSample, stopSample, looping ? 1 : 0) != 0;

  void stop() => _stop();
  bool get isPlaying => _isPlaying() != 0;

  /// Samples handed to the device — **the clock**. The picture reads this
  /// and shows whatever frame it lands in; when rendering falls behind,
  /// frames are dropped rather than the sound being made to wait.
  int get positionSamples => _position();

  /// Moves the transport without restarting anything. Because the mixer
  /// builds a mix rather than starting clips, a seek is just a change of
  /// where the next block is read from.
  void seek(int sample) => _seek(sample);

  /// The last mixed block's PRE-CLIP bus peak (0 = left, 1 = right; mono
  /// mirrors left) — the level meter's feed (AUDIO-PRO R2). A value past
  /// 1.0 means the output stage is clipping, which is exactly what the
  /// meter exists to make visible.
  double peakFor(int channel) => _peak(channel);

  /// Opens the capture device and starts delivering into the C-side ring
  /// (AUDIO-PRO R5). Returns the delivered sample rate ([sampleRate] — the
  /// C converts from the device's native rate) or 0 on failure, which on
  /// mobile/macOS includes "no microphone permission". Channels are the
  /// device's own — read [captureChannels] after a successful start.
  ///
  /// Independent of the playback device: recording runs DURING playback.
  int captureStart({
    required int sampleRate,
    bool useNullBackend = false,
    int deviceIndex = -1,
  }) => _captureStart(sampleRate, useNullBackend ? 1 : 0, deviceIndex);

  /// Drains up to [maxFloats] captured floats into [out], returning how
  /// many were copied. Call from a timer while recording, and once more
  /// before [captureStop] for the tail.
  int captureRead(Pointer<Float> out, int maxFloats) =>
      _captureRead(out, maxFloats);

  void captureStop() => _captureStop();
  bool get captureIsOpen => _captureIsOpen() != 0;
  int get captureChannels => _captureChannels();

  /// Frames the ring dropped because the drain fell behind. Nonzero means
  /// the take is damaged — say so, never save a silently shortened file.
  int get captureDroppedFrames => _captureDroppedFrames();

  /// The capture path's own buffering in samples (typically 10-30 ms).
  int get captureLatencySamples => _captureLatency();
}

/// The enumeration index for the output device named [name], or -1 (the
/// system default) when [name] is null or no longer attached — a missing
/// speaker falls back to the default rather than failing playback
/// (AUDIO-PRO R4).
int audioOutputDeviceIndexByName(QaAudioDevice device, String? name) {
  if (name == null) {
    return -1;
  }
  final devices = device.devicesOf(capture: false);
  for (var index = 0; index < devices.length; index += 1) {
    if (devices[index].name == name) {
      return index;
    }
  }
  return -1;
}

/// The capture-side twin: the input device named [name], or -1 (the system
/// default microphone) when [name] is null or unplugged (AUDIO-PRO R5).
int audioInputDeviceIndexByName(QaAudioDevice device, String? name) {
  if (name == null) {
    return -1;
  }
  final devices = device.devicesOf(capture: true);
  for (var index = 0; index < devices.length; index += 1) {
    if (devices[index].name == name) {
      return index;
    }
  }
  return -1;
}

/// Reads the played position as a frame index, pulled forward by the
/// device's own reported latency so the picture matches what is being
/// HEARD rather than what has merely been queued.
///
/// [extraOffsetSamples] is the user's A/V offset: the residual no device
/// report can account for (screen pipeline, Bluetooth, an AV receiver).
/// Every professional tool exposes this for the same reason — the part
/// that is measurable gets corrected automatically, and the rest is a
/// slider.
int audioClockFrame({
  required int positionSamples,
  required int latencySamples,
  required int extraOffsetSamples,
  required int sampleRate,
  required int frameRateNumerator,
  required int frameRateDenominator,
}) {
  if (sampleRate <= 0 || frameRateNumerator <= 0 || frameRateDenominator <= 0) {
    return 0;
  }
  final heard = positionSamples - latencySamples + extraOffsetSamples;
  if (heard <= 0) {
    return 0;
  }
  // Integer throughout, like every other frame/sample conversion in this
  // program: a double here would drift over a long timeline.
  return heard * frameRateNumerator ~/ (sampleRate * frameRateDenominator);
}
